sasmodels/models/lib/magnetic_functions.c

```diff
@@ -0,0 +1,193 @@
+static double clipp(double value, double low, double high) //from kernel_iq.c
+{
+  return (value < low ? low : (value > high ? high : value));
+} 
+
+static double length(double x, double y) 
+{
+  return sqrt(x*x + y*y);
+} 
+
+static double fq_core_shell(double q, double sld_core, double radius,
+   double sld_solvent, double fp_n, double sld[], double thickness[])
+{
+  const int n = (int)(fp_n+0.5);
+  double f, r, last_sld;
+  r = radius;
+  last_sld = sld_core;
+  f = 0.;
+  for (int i=0; i<n; i++) {
+    f += M_4PI_3 * cube(r) * (sld[i] - last_sld) * sas_3j1x_x(q*r);
+    last_sld = sld[i];
+    r += thickness[i];
+  }
+  f += M_4PI_3 * cube(r) * (sld_solvent - last_sld) * sas_3j1x_x(q*r);
+  return f;
+}
+
+static double langevin(
+    double x) {
+    // cotanh(x)-1\x
+
+    if (x < 0.00001) {
+        // avoid dividing by zero
+        return 1.0/3.0*x;
+    } else {
+        return 1.0/tanh(x)-1/x;
+    }
+}
+
+static double langevinoverx(
+    double x) 
+{
+    // cotanh(x)-1\x
+
+    if (x < 0.00001) {
+        // avoid dividing by zero
+        return 1.0/3.0;
+    } else {
+        return langevin(x)/x;
+    }
+}
+
```

Comment:
Consider moving these functions into lib and including them automatically in the magnetic branch of `kernel_iq.c`.

This is not a drop-in replacement. This code defines ++ and -- imaginary sld terms which are not defined in the kernel_iq.c.

A more radical approach is to drop automatic magnetic models from sasview/sasmodels. The fact that we need a specialized model for a core-shell magnetic form factor suggests that the generic forms are of no use to anybody.

Has anyone ever used sasview magnetic models to fit data?